File: frontend/lib/api/conversations.ts
/**
 * API client for the conversations domain.
 * Calls FastAPI endpoints defined in backend/app/routers/conversations.py.
 */

const API_URL = process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:8000';

export interface Contact {
  id: string;
  phone: string;
  name: string | null;
  email: string | null;
  tags: string[];
  notes: string | null;
  is_blocked: boolean;
  created_at: string;
  updated_at: string;
}

export interface Conversation {
  id: string;
  contact_id: string;
  contact_phone: string;
  status: 'bot' | 'open' | 'closed';
  unread_count: number;
  last_message_at: string | null;
  last_message: string | null;
  assigned_to: string | null;
  created_at: string;
  updated_at: string;
  contact: Contact | null;
}

export interface Message {
  id: string;
  conversation_id: string;
  direction: 'inbound' | 'outbound';
  type: string;
  body: string | null;
  media_url: string | null;
  status: 'pending' | 'sent' | 'delivered' | 'read' | 'failed' | 'received';
  created_at: string;
}

export interface ConversationListResponse {
  items: Conversation[];
  total: number;
  page: number;
  page_size: number;
}

export interface ConversationThreadResponse {
  conversation: Conversation;
  messages: Message[];
  total_messages: number;
}

async function apiFetch<T>(path: string, token: string, options: RequestInit = {}): Promise<T> {
  const res = await fetch(`${API_URL}${path}`, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
      ...(options.headers as Record<string, string> | undefined),
    },
  });

  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body?.error ?? body?.detail ?? `HTTP ${res.status}`);
  }

  return res.json() as Promise<T>;
}

export async function listConversations(
  token: string,
  params?: { page?: number; page_size?: number; status?: string },
): Promise<ConversationListResponse> {
  const query = new URLSearchParams();
  if (params?.page) query.set('page', String(params.page));
  if (params?.page_size) query.set('page_size', String(params.page_size));
  if (params?.status) query.set('status', params.status);
  const qs = query.toString() ? `?${query.toString()}` : '';
  return apiFetch<ConversationListResponse>(`/api/v1/conversations${qs}`, token);
}

export async function getConversationThread(
  token: string,
  conversationId: string,
): Promise<ConversationThreadResponse> {
  return apiFetch<ConversationThreadResponse>(
    `/api/v1/conversations/${conversationId}`,
    token,
  );
}

export async function updateConversation(
  token: string,
  conversationId: string,
  data: { status?: string; assigned_to?: string },
): Promise<Conversation> {
  return apiFetch<Conversation>(`/api/v1/conversations/${conversationId}`, token, {
    method: 'PATCH',
    body: JSON.stringify(data),
  });
}

File: backend/app/models/conversations.py
from __future__ import annotations
from uuid import UUID
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict


# ── Contact ───────────────────────────────────────────────────────────────────

class ContactResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    phone: str
    name: Optional[str]
    email: Optional[str]
    tags: list[str]
    notes: Optional[str]
    is_blocked: bool
    created_at: datetime
    updated_at: datetime


# ── Message ───────────────────────────────────────────────────────────────────

class MessageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    conversation_id: UUID
    direction: str          # "inbound" | "outbound"
    type: str
    body: Optional[str]
    media_url: Optional[str]
    status: str             # "pending" | "sent" | "delivered" | "read" | "failed"
    created_at: datetime


# ── Conversation ──────────────────────────────────────────────────────────────

class ConversationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    contact_id: UUID
    contact_phone: str
    status: str             # "bot" | "open" | "closed"
    unread_count: int
    last_message_at: Optional[datetime]
    last_message: Optional[str]
    assigned_to: Optional[UUID]
    created_at: datetime
    updated_at: datetime
    contact: Optional[ContactResponse] = None


class ConversationListResponse(BaseModel):
    items: list[ConversationResponse]
    total: int
    page: int
    page_size: int


class ConversationThreadResponse(BaseModel):
    conversation: ConversationResponse
    messages: list[MessageResponse]
    total_messages: int


class UpdateConversationRequest(BaseModel):
    status: Optional[str] = None    # "bot" | "open" | "closed"
    assigned_to: Optional[UUID] = None
